File: Python/CodeSignal/Industry Framework/File Storage System/level3.py
class CloudStorage:
    def __init__(self):
        self.store = {} # name -> (size, owner)
        self.users = {} # userid -> capacity_limit
        self.user_usage = {} # userid -> current_usage
    
    def add_file(self, name, size):
        if name in self.store:
            return False
        
        # Admin has unlimited capacity, all files are owned by admin
        self.store[name] = (size, "admin")
        return True
    
    def get_file_size(self, name):
        if name not in self.store:
            return ""
        
        return self.store[name][0]
    
    def delete_file(self, name):
        if name not in self.store:
            return ""
        
        size, owner = self.store[name]
        del self.store[name]
        
        # Update user usage if not admin
        if owner != "admin" and owner in self.user_usage:
            self.user_usage[owner] -= int(size)
        
        return size
    
    def get_n_largest(self, prefix, n):
        matching_files = []
        for name, (size, owner) in self.store.items():
            if name.startswith(prefix):
                matching_files.append((name, int(size)))
        
        if not matching_files:
            return ""
        
        matching_files.sort(key=lambda x: (-x[1], x[0]))
        result_files = matching_files[:n]
        result_parts = [f"{name}({size})" for name, size in result_files]
        return ", ".join(result_parts)
    
    def add_user(self, userid, capacity):
        if userid in self.users:
            return False
        
        self.users[userid] = capacity
        self.user_usage[userid] = 0
        return True
    
    def add_file_by(self, userid, name, size):
        # Check if file already exists or user doesn't exist
        if name in self.store or userid not in self.users:
            return ""
        
        # Check if user has enough capacity
        if self.user_usage[userid] + size > self.users[userid]:
            return ""
        
        # Add file and update usage
        self.store[name] = (str(size), userid)
        self.user_usage[userid] += size
        
        # Return remaining capacity
        return str(self.users[userid] - self.user_usage[userid])
    
    def merge_user(self, userid1, userid2):
        # Check if both users exist and are different
        if userid1 not in self.users or userid2 not in self.users or userid1 == userid2:
            return ""
        
        # Transfer all files from userid2 to userid1
        for name, (size, owner) in list(self.store.items()):
            if owner == userid2:
                self.store[name] = (size, userid1)
        
        # Merge capacities and usage
        self.users[userid1] += self.users[userid2]
        self.user_usage[userid1] += self.user_usage[userid2]
        
        # Remove userid2
        del self.users[userid2]
        del self.user_usage[userid2]
        
        # Return remaining capacity of userid1
        return str(self.users[userid1] - self.user_usage[userid1])

def main():
    storage = CloudStorage()
    
    # Test cases from level 3 example
    queries = [
        ["ADD_USER", "user1", "200"],
        ["ADD_USER", "user1", "100"],
        ["ADD_FILE_BY", "user1", "/dir/file.med", "50"],
        ["ADD_FILE_BY", "user1", "/file.big", "140"],
        ["ADD_FILE_BY", "user1", "/dir/file.small", "20"],
        ["ADD_FILE", "/admin_file", "300"],
        ["ADD_USER", "user2", "110"],
        ["ADD_FILE_BY", "user2", "/dir/file.med", "45"],
        ["ADD_FILE_BY", "user2", "/new_file", "90"],
        ["MERGE_USER", "user1", "user2"]
    ]
    
    results = []
    for query in queries:
        operation = query[0]
        if operation == "ADD_FILE":
            result = storage.add_file(query[1], query[2])
            results.append(str(result).lower())
        elif operation == "GET_FILE_SIZE":
            result = storage.get_file_size(query[1])
            results.append(result)
        elif operation == "DELETE_FILE":
            result = storage.delete_file(query[1])
            results.append(result)
        elif operation == "GET_N_LARGEST":
            result = storage.get_n_largest(query[1], int(query[2]))
            results.append(result)
        elif operation == "ADD_USER":
            result = storage.add_user(query[1], int(query[2]))
            results.append(str(result).lower())
        elif operation == "ADD_FILE_BY":
            result = storage.add_file_by(query[1], query[2], int(query[3]))
            results.append(result)
        elif operation == "MERGE_USER":
            result = storage.merge_user(query[1], query[2])
            results.append(result)
    
    print(results)
    # Expected output: ["true", "false", "150", "10", "", "true", "true", "", "60", "70"]

if __name__ == "__main__":
    main()

File: Java/A-PREP/Code Signal Industry/Cloud Storage System/level3.java
import java.util.*;

public class CloudStorageSystem {
    // Map to store file names and their sizes
    private Map<String, Integer> files;
    // Map to store file ownership (file name -> user ID)
    private Map<String, String> fileOwners;
    // Map to store users and their storage capacities
    private Map<String, Integer> users;

    // Constructor to initialize the maps
    public CloudStorageSystem() {
        files = new HashMap<>();
        fileOwners = new HashMap<>();
        users = new HashMap<>();
        // Add the admin user with unlimited capacity
        users.put("admin", Integer.MAX_VALUE);
    }

    // Method to add a new user
    public boolean addUser(String userId, int capacity) {
        // Check if the user already exists
        if (users.containsKey(userId)) {
            return false; // User already exists
        }
        // Add the user with the given capacity
        users.put(userId, capacity);
        return true; // User added successfully
    }

    // Method to add a new file by a specific user
    public String addFileByUser(String userId, String name, int size) {
        // Check if the user exists
        if (!users.containsKey(userId)) {
            return ""; // User does not exist
        }
        // Check if the file already exists
        if (files.containsKey(name)) {
            return ""; // File already exists
        }
        // Check if the user has enough remaining capacity
        int remainingCapacity = users.get(userId);
        if (remainingCapacity < size) {
            return ""; // Not enough capacity
        }
        // Add the file and update the user's remaining capacity
        files.put(name, size);
        fileOwners.put(name, userId);
        users.put(userId, remainingCapacity - size);
        return String.valueOf(remainingCapacity - size); // Return the remaining capacity
    }

    // Method to add a new file (admin operation)
    public boolean addFile(String name, int size) {
        // Check if the file already exists
        if (files.containsKey(name)) {
            return false; // File already exists
        }
        // Add the file (admin has unlimited capacity)
        files.put(name, size);
        fileOwners.put(name, "admin");
        return true; // File added successfully
    }

    // Method to merge two user accounts
    public String mergeUsers(String userId1, String userId2) {
        // Check if both users exist and are different
        if (!users.containsKey(userId1) || !users.containsKey(userId2) || userId1.equals(userId2)) {
            return ""; // Invalid merge
        }
        // Transfer ownership of all files from userId2 to userId1
        for (Map.Entry<String, String> entry : fileOwners.entrySet()) {
            if (entry.getValue().equals(userId2)) {
                fileOwners.put(entry.getKey(), userId1);
            }
        }
        // Add userId2's remaining capacity to userId1
        int remainingCapacity1 = users.get(userId1);
        int remainingCapacity2 = users.get(userId2);
        users.put(userId1, remainingCapacity1 + remainingCapacity2);
        // Remove userId2 from the system
        users.remove(userId2);
        return String.valueOf(remainingCapacity1 + remainingCapacity2); // Return the new remaining capacity
    }

    // Main method to test the functionality
    public static void main(String[] args) {
        // Create a new cloud storage system
        CloudStorageSystem storage = new CloudStorageSystem();

        // Test case 1: Add a new user
        System.out.println(storage.addUser("user1", 200)); // true

        // Test case 2: Try to add the same user again
        System.out.println(storage.addUser("user1", 300)); // false

        // Test case 3: Add a file by user1
        System.out.println(storage.addFileByUser("user1", "/dir/file.mad", 50)); // "150"

        // Test case 4: Add another file by user1
        System.out.println(storage.addFileByUser("user1", "/file.hig", 140)); // "10"

        // Test case 5: Try to add a file that exceeds user1's capacity
        System.out.println(storage.addFileByUser("user1", "/dir/file.email", 20)); // ""

        // Test case 6: Add a file by admin (unlimited capacity)
        System.out.println(storage.addFile("/dir/saint_file", 300)); // true

        // Test case 7: Add a new user
        System.out.println(storage.addUser("user2", 110)); // true

        // Test case 8: Try to add a file that already exists
        System.out.println(storage.addFileByUser("user2", "/dir/file.mad", 45)); // ""

        // Test case 9: Add a file by user2
        System.out.println(storage.addFileByUser("user2", "/new_file", 50)); // "60"

        // Test case 10: Merge user1 and user2
        System.out.println(storage.mergeUsers("user1", "user2")); // "70"
    }
}